File: libs/src/java/salmon-fs/com/mku/salmonfs/utils/AesFileComparators.java
package com.mku.salmonfs.utils;
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import com.mku.salmonfs.file.AesFile;
import com.mku.fs.drive.utils.FileUtils;

import java.util.Comparator;

/**
 * Useful comparators for SalmonFile.
 */
public class AesFileComparators {
    private static final Comparator<AesFile> defaultComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return -1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return 1;
        else return 0;
    };

    private static final Comparator<AesFile> filenameAscComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return -1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return 1;
        else
            return tryGetBasename(c1).compareTo(tryGetBasename(c2));
    };

    private static final Comparator<AesFile> filenameDescComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return 1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return -1;
        else
            return tryGetBasename(c2).compareTo(tryGetBasename(c1));
    };

    private static final Comparator<AesFile> sizeAscComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return -1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return 1;
        else
            return (int) (tryGetSize(c1) - tryGetSize(c2));
    };

    private static final Comparator<AesFile> sizeDescComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return 1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return -1;
        else
            return (int) (tryGetSize(c2) - tryGetSize(c1));
    };

    private static final Comparator<AesFile> typeAscComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return -1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return 1;
        else
            return tryGetType(c1).compareTo(tryGetType(c2));
    };

    private static final Comparator<AesFile> typeDescComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return 1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return -1;
        else
            return tryGetType(c2).compareTo(tryGetType(c1));
    };

    private static final Comparator<AesFile> dateAscComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return -1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return 1;
        else
            return (int) (tryGetDate(c1) - tryGetDate(c2));
    };

    private static final Comparator<AesFile> dateDescComparator = (AesFile c1, AesFile c2) ->
    {
        if (c1.isDirectory() && !c2.isDirectory())
            return 1;
        else if (!c1.isDirectory() && c2.isDirectory())
            return -1;
        else
            return (int) (tryGetDate(c2) - tryGetDate(c1));
    };

    private static final Comparator<AesFile> relevanceComparator = (AesFile c1, AesFile c2) ->
            (int) c2.getTag() - (int) c1.getTag();

    /**
     * Get default comparator. This is the fastest sorting comparator as it simply lists folders first.
     * The rest of the files will be listed without sorting.
     * @return The default comparator
     */
    public static Comparator<AesFile> getDefaultComparator() {
        return defaultComparator;
    }

    /**
     * Get Filename Ascending Comparator.
     * @return The filename scending comparator.
     */
    public static Comparator<AesFile> getFilenameAscComparator() {
        return filenameAscComparator;
    }

    /**
     * Get Filename Descending Comparator.
     * @return filename descending comparator
     */
    public static Comparator<AesFile> getFilenameDescComparator() {
        return filenameDescComparator;
    }

    /**
     * Get Size Ascending Comparator.
     * @return The size ascending comparator
     */
    public static Comparator<AesFile> getSizeAscComparator() {
        return sizeAscComparator;
    }

    /**
     * Get Size Descending Comparator.
     * @return The size descending comparator
     */
    public static Comparator<AesFile> getSizeDescComparator() {
        return sizeDescComparator;
    }

    /**
     * Get File Type Ascending Comparator.
     * @return The file type ascending comparator
     */
    public static Comparator<AesFile> getTypeAscComparator() {
        return typeAscComparator;
    }

    /**
     * Get File Type Descending Comparator.
     * @return The file type descending comparator
     */
    public static Comparator<AesFile> getTypeDescComparator() {
        return typeDescComparator;
    }

    /**
     * Get Date Ascending Comparator.
     * @return The date ascending comparator
     */
    public static Comparator<AesFile> getDateAscComparator() {
        return dateAscComparator;
    }

    /**
     * Get Date Descending Comparator.
     * @return The date descending comparator
     */
    public static Comparator<AesFile> getDateDescComparator() {
        return dateDescComparator;
    }

    /**
     * Get Relevant Comparator. This will sort by the search result relevancy in the Tag member. Most relevant results
     * will be listed first.
     * @return The relevancy comparator
     */
    public static Comparator<AesFile> getRelevanceComparator() {
        return relevanceComparator;
    }

    /**
     * Get the SalmonFile basename if available.
     * @param aesFile The file
     * @return The base name
     */
    private static String tryGetBasename(AesFile aesFile) {
        try {
            return aesFile.getBaseName();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return "";
    }

    /**
     * Get the SalmonFile file type extension if available.
     * @param aesFile The Aes file to get the file type from
     * @return The file type
     */
    private static String tryGetType(AesFile aesFile) {
        try {
            if (aesFile.isDirectory())
                return aesFile.getBaseName();
            return FileUtils.getExtensionFromFileName(aesFile.getBaseName()).toLowerCase();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return "";
    }

    /**
     * Get the SalmonFile size if available.
     * @param aesFile The file
     * @return The size
     */
    private static long tryGetSize(AesFile aesFile) {
        try {
            if (aesFile.isDirectory())
                return aesFile.getChildrenCount();
            // the original file length requires reading the chunks size
            // from the file so it can get a bit expensive
            // so instead we sort on the real file size
            return aesFile.getRealFile().length();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return 0;
    }

    /**
     * Get the SalmonFile date if available.
     * @param aesFile The file
     * @return The date in milliseconds
     */
    private static long tryGetDate(AesFile aesFile) {
        try {
            return aesFile.getLastDateTimeModified();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return 0;
    }
}

File: libs/src/csharp/SalmonFS/SalmonFS/Drive/Utils/AesFileComparators.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

using Mku.FS.Drive.Utils;
using Mku.SalmonFS.File;
using System;
using System.Collections.Generic;

namespace Mku.SalmonFS.Drive.Utils;

/// <summary>
///  Useful Comparers for AesFile.
/// </summary>
public class AesFileComparators
{

    /// <summary>
    ///  Get default Comparer. This is the fastest sorting Comparer as it simply lists folders first.
    ///  The rest of the files will be listed without sorting.
    /// </summary>
    ///  <returns>The default comparer</returns>
    public static Comparer<AesFile> DefaultComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return -1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return 1;
        else return 0;
    });

    /// <summary>
    ///  Get Filename Ascending Comparer.
	/// </summary>
	///  <returns>The filename ascending comparer</returns>
    public static Comparer<AesFile> FilenameAscComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return -1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return 1;
        else
            return TryGetBasename(c1).CompareTo(TryGetBasename(c2));
    });

    /// <summary>
    ///  Get Filename Descending Comparer.
	/// </summary>
	///  <returns>The filename descending comparer</returns>
    public static Comparer<AesFile> FilenameDescComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return 1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return -1;
        else
            return TryGetBasename(c2).CompareTo(TryGetBasename(c1));
    });

    /// <summary>
    ///  Get Size Ascending Comparer.
	/// </summary>
	///  <returns>The size ascending comparer</returns>
    public static Comparer<AesFile> SizeAscComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return -1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return 1;
        else
            return (int)(TryGetSize(c1) - TryGetSize(c2));
    });

    /// <summary>
    ///  Get Size Descending Comparer.
	/// </summary>
	///  <returns>The size descending comparer</returns>
    public static Comparer<AesFile> SizeDescComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return 1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return -1;
        else
            return (int)(TryGetSize(c2) - TryGetSize(c1));
    });

    /// <summary>
    ///  Get File Type Ascending Comparer.
	/// </summary>
	///  <returns>The file type ascending comparer</returns>
    public static Comparer<AesFile> TypeAscComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return -1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return 1;
        else
            return TryGetType(c1).CompareTo(TryGetType(c2));
    });

    /// <summary>
    ///  Get Filetype Descending Comparer.
	/// </summary>
	///  <returns>The file type descending comparer</returns>
    public static Comparer<AesFile> TypeDescComparator { get; } =
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return 1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return -1;
        else
            return TryGetType(c2).CompareTo(TryGetType(c1));
    });

    /// <summary>
    ///  Get Date Ascending Comparer.
	/// </summary>
	///  <returns>The date ascending comparer</returns>
    public static Comparer<AesFile> DateAscComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return -1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return 1;
        else
            return (int)(TryGetDate(c1) - TryGetDate(c2));
    });

    /// <summary>
    ///  Get Date Descending Comparer.
	/// </summary>
	///  <returns>The date descending comparer</returns>
    public static Comparer<AesFile> DateDescComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
    {
        if (c1.IsDirectory && !c2.IsDirectory)
            return 1;
        else if (!c1.IsDirectory && c2.IsDirectory)
            return -1;
        else
            return (int)(TryGetDate(c2) - TryGetDate(c1));
    });

    /// <summary>
    ///  Get Relevant Comparer. This will sort by the search result relevancy in the Tag member. Most relevant results
    ///  will be listed first.
	/// </summary>
	///  <returns>The relevance comparer</returns>
    public static Comparer<AesFile> RelevanceComparator { get; } = 
        Comparer<AesFile>.Create((AesFile c1, AesFile c2) =>
            (int)c2.Tag - (int)c1.Tag);


    /// <summary>
    ///  Get the AesFile base name if available.
    /// </summary>
    ///  <param name="salmonFile">The file</param>
    ///  <returns>The base name</returns>
    private static string TryGetBasename(AesFile salmonFile)
    {
        try
        {
            return salmonFile.Name;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
        }
        return "";
    }

    /// <summary>
    ///  Get the AesFile file type extension if available.
	/// </summary>
	///  <param name="salmonFile">The file</param>
    ///  <returns>The file type</returns>
    private static string TryGetType(AesFile salmonFile)
    {
        try
        {
            if (salmonFile.IsDirectory)
                return salmonFile.Name;
            return FileUtils.GetExtensionFromFileName(salmonFile.Name).ToLower();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
        }
        return "";
    }

    /// <summary>
    ///  Get the AesFile size if available.
	/// </summary>
	///  <param name="salmonFile">The file</param>
    ///  <returns>The file size</returns>
    private static long TryGetSize(AesFile salmonFile)
    {
        try
        {
            if (salmonFile.IsDirectory)
                return salmonFile.ChildrenCount;
            // the original file length requires reading the chunks size
            // from the file so it can get a bit expensive
            // so instead we sort on the real file size
            return salmonFile.RealFile.Length;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
        }
        return 0;
    }

    /// <summary>
    ///  Get the AesFile date if available.
	/// </summary>
	///  <param name="salmonFile">The file</param>
    ///  <returns>The file date in milliseconds</returns>
    private static long TryGetDate(AesFile salmonFile)
    {
        try
        {
            return salmonFile.LastDateModified;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
        }
        return 0;
    }
}